src/interfaces/messages/handlers/messages-get.ts

```diff
@@ -0,0 +1,89 @@
+import type { DataStore } from '../../../store/data-store.js';
+import type { DidResolver } from '../../../did/did-resolver.js';
+import type { MessageStore } from '../../../store/message-store.js';
+import type { MethodHandler } from '../../types.js';
+import type { MessagesGetMessage, MessagesGetReply, MessagesGetReplyEntry } from '../types.js';
+
+import { DataStream } from '../../../utils/data-stream.js';
+import { DwnConstant } from '../../../core/dwn-constant.js';
+import { Encoder } from '../../../utils/encoder.js';
+import { MessageReply } from '../../../core/message-reply.js';
+import { MessagesGet } from '../messages/messages-get.js';
+import { authenticate, authorize } from '../../../core/auth.js';
+import { DwnInterfaceName, DwnMethodName, Message } from '../../../core/message.js';
+
+type HandleArgs = { tenant: string, message: MessagesGetMessage };
+
+export class MessagesGetHandler implements MethodHandler {
+  constructor(private didResolver: DidResolver, private messageStore: MessageStore, private dataStore: DataStore) {}
+
+  public async handle({ tenant, message }: HandleArgs): Promise<MessagesGetReply> {
+    let messagesGet: MessagesGet;
+
+    try {
+      messagesGet = await MessagesGet.parse(message);
+    } catch (e) {
+      return MessageReply.fromError(e, 400);
+    }
+
+    try {
+      await authenticate(message.authorization, this.didResolver);
+      await authorize(tenant, messagesGet);
+    } catch (e) {
+      return MessageReply.fromError(e, 401);
+    }
+
+    const promises: Promise<MessagesGetReplyEntry>[] = [];
+
+    for (const messageCid of message.descriptor.messageCids) {
+      const promise = this.messageStore.get(tenant, messageCid)
+        .then(message => {
+          return { messageCid, message };
+        })
+        .catch(_ => {
```

Comment:
i dont think `MessageStore` will ever throw (unless the operation is aborted).  you probably should check for `message === undefined` instead

Reply:
`message-store-level` should not throw, but it's possible a custom message store will. In that case, it's useful to `catch`. Also, don't think we should throw/log error on `message === undefined`. That's a non-exceptional case IMO, and the client should be okay with receiving `message: undefined` in the reply entry.

Reply:
> `message-store-level` should not throw, but it's possible a custom message store will. In that case, it's useful to `catch`.  Also, don't think we should throw/log error on `message === undefined`. That's a non-exceptional case IMO, and the client should be okay with receiving `message: undefined` in the reply entry.

💯